diff --git a/anyUtil/any_to_float_x_test.go b/anyUtil/any_to_float_x_test.go
index 3d0a55e..8cc7669 100644
--- a/anyUtil/any_to_float_x_test.go
+++ b/anyUtil/any_to_float_x_test.go
@@ -1,7 +1,6 @@
 package anyUtil
 
 import (
-	go_easy_utils "github.com/jefferyjob/go-easy-utils/v2"
 	"math"
 	"testing"
 )
@@ -30,8 +29,8 @@ func TestAnyToFloat32(t *testing.T) {
 		{uint32(32), 32, nil},
 		{uint64(64), 64, nil},
 		{"3.14", 3.14, nil},
-		{math.MaxFloat64, 0, go_easy_utils.ErrValOut},
-		{make(chan int), 0, go_easy_utils.ErrType},
+		{math.MaxFloat64, 0, ErrValOut},
+		{make(chan int), 0, ErrType},
 	}
 
 	for _, test := range tests {
@@ -79,8 +78,8 @@ func TestAnyToFloat64(t *testing.T) {
 		{complex128(1 + 2i), 1, nil},
 		{&iPtr, 90, nil},
 		{(*int)(nil), 0, nil},
-		{"abc", 0, go_easy_utils.ErrSyntax},
-		{[]int{}, 0, go_easy_utils.ErrType},
+		{"abc", 0, ErrSyntax},
+		{[]int{}, 0, ErrType},
 	}
 
 	for _, test := range tests {
diff --git a/anyUtil/any_to_int_x_test.go b/anyUtil/any_to_int_x_test.go
index 6c75509..e7405d5 100644
--- a/anyUtil/any_to_int_x_test.go
+++ b/anyUtil/any_to_int_x_test.go
@@ -2,7 +2,6 @@ package anyUtil
 
 import (
 	"errors"
-	"github.com/jefferyjob/go-easy-utils/v2"
 	"math"
 	"testing"
 )
@@ -32,7 +31,7 @@ func TestAnyToInt(t *testing.T) {
 		{input: float32(123.45), expected: 123, err: nil},
 		{input: float64(123.45), expected: 123, err: nil},
 		{input: "123", expected: 123, err: nil},
-		{input: "abc", expected: 0, err: go_easy_utils.ErrSyntax},
+		{input: "abc", expected: 0, err: ErrSyntax},
 	}
 
 	for _, tc := range testCases {
@@ -83,13 +82,13 @@ func TestAnyToInt8(t *testing.T) {
 			name:     "positive integer out of range",
 			input:    128,
 			expected: 0,
-			err:      go_easy_utils.ErrValOut,
+			err:      ErrValOut,
 		},
 		{
 			name:     "negative integer out of range",
 			input:    -129,
 			expected: 0,
-			err:      go_easy_utils.ErrValOut,
+			err:      ErrValOut,
 		},
 		{
 			name:     "float",
@@ -107,13 +106,13 @@ func TestAnyToInt8(t *testing.T) {
 			name:     "string out of range",
 			input:    "128",
 			expected: 0,
-			err:      go_easy_utils.ErrValOut,
+			err:      ErrValOut,
 		},
 		{
 			name:     "invalid string",
 			input:    "invalid",
 			expected: 0,
-			err:      go_easy_utils.ErrSyntax,
+			err:      ErrSyntax,
 		},
 	}
 
@@ -159,8 +158,8 @@ func TestAnyToInt16(t *testing.T) {
 		{float32(123.45), 123, nil},
 		{float64(123.45), 123, nil},
 		{"12345", 12345, nil},
-		{math.MinInt16 - 1, 0, go_easy_utils.ErrValOut},
-		{"not a number", 0, go_easy_utils.ErrSyntax},
+		{math.MinInt16 - 1, 0, ErrValOut},
+		{"not a number", 0, ErrSyntax},
 	}
 
 	for _, tt := range tests {
@@ -184,13 +183,13 @@ func TestAnyToInt32(t *testing.T) {
 	}{
 		{int(123), 123, nil},
 		{int64(2147483647), 2147483647, nil},
-		{int64(2147483648), 0, go_easy_utils.ErrValOut},
+		{int64(2147483648), 0, ErrValOut},
 		{int64(-2147483648), -2147483648, nil},
-		{int64(-2147483649), 0, go_easy_utils.ErrValOut},
+		{int64(-2147483649), 0, ErrValOut},
 		{float64(123.45), 123, nil},
 		{"123", 123, nil},
-		{"-2147483649", 0, go_easy_utils.ErrValOut},
-		{struct{}{}, 0, go_easy_utils.ErrType},
+		{"-2147483649", 0, ErrValOut},
+		{struct{}{}, 0, ErrType},
 	}
 
 	for _, testCase := range testCases {
diff --git a/anyUtil/any_to_uint_x_test.go b/anyUtil/any_to_uint_x_test.go
index ae47560..09f80d7 100644
--- a/anyUtil/any_to_uint_x_test.go
+++ b/anyUtil/any_to_uint_x_test.go
@@ -1,7 +1,6 @@
 package anyUtil
 
 import (
-	"github.com/jefferyjob/go-easy-utils/v2"
 	"reflect"
 	"testing"
 )
@@ -14,11 +13,11 @@ func TestAnyToUint(t *testing.T) {
 		err    error
 	}{
 		{10, 10, nil},
-		{-5, 0, go_easy_utils.ErrUnsignedInt},
+		{-5, 0, ErrUnsignedInt},
 		{"20", 20, nil},
 		{1.5, 1, nil},
 		{2.5, 2, nil},
-		{make(chan int), 0, go_easy_utils.ErrType},
+		{make(chan int), 0, ErrType},
 	}
 
 	// Test loop
@@ -39,10 +38,10 @@ func TestAnyToUint8(t *testing.T) {
 		err    error
 	}{
 		{10, 10, nil},
-		{300, 0, go_easy_utils.ErrValOut},
+		{300, 0, ErrValOut},
 		{"20", 20, nil},
 		{1.5, 1, nil},
-		{make(chan int), 0, go_easy_utils.ErrType},
+		{make(chan int), 0, ErrType},
 	}
 
 	// Test loop
@@ -63,10 +62,10 @@ func TestAnyToUint16(t *testing.T) {
 		err    error
 	}{
 		{10, 10, nil},
-		{70000, 0, go_easy_utils.ErrValOut},
+		{70000, 0, ErrValOut},
 		{"20", 20, nil},
 		{1.5, 1, nil},
-		{make(chan int), 0, go_easy_utils.ErrType},
+		{make(chan int), 0, ErrType},
 	}
 
 	// Test loop
@@ -87,10 +86,10 @@ func TestAnyToUint32(t *testing.T) {
 		err    error
 	}{
 		{10, 10, nil},
-		{5000000000, 0, go_easy_utils.ErrValOut},
+		{5000000000, 0, ErrValOut},
 		{"20", 20, nil},
 		{1.5, 1, nil},
-		{make(chan int), 0, go_easy_utils.ErrType},
+		{make(chan int), 0, ErrType},
 	}
 
 	// Test loop
@@ -115,37 +114,37 @@ func TestAnyToUint64(t *testing.T) {
 		{
 			name:      "Test -float32",
 			input:     float32(-0.1),
-			wantError: go_easy_utils.ErrUnsignedInt,
+			wantError: ErrUnsignedInt,
 		},
 		{
 			name:      "Test -float64",
 			input:     float64(-0.2),
-			wantError: go_easy_utils.ErrUnsignedInt,
+			wantError: ErrUnsignedInt,
 		},
 		{
 			name:      "Test -int",
 			input:     int(-1),
-			wantError: go_easy_utils.ErrUnsignedInt,
+			wantError: ErrUnsignedInt,
 		},
 		{
 			name:      "Test -int8",
 			input:     int8(-2),
-			wantError: go_easy_utils.ErrUnsignedInt,
+			wantError: ErrUnsignedInt,
 		},
 		{
 			name:      "Test -int16",
 			input:     int16(-3),
-			wantError: go_easy_utils.ErrUnsignedInt,
+			wantError: ErrUnsignedInt,
 		},
 		{
 			name:      "Test -int32",
 			input:     int32(-4),
-			wantError: go_easy_utils.ErrUnsignedInt,
+			wantError: ErrUnsignedInt,
 		},
 		{
 			name:      "Test -int64",
 			input:     int64(-5),
-			wantError: go_easy_utils.ErrUnsignedInt,
+			wantError: ErrUnsignedInt,
 		},
 		{
 			name:  "Test uint",
@@ -220,7 +219,7 @@ func TestAnyToUint64(t *testing.T) {
 		{
 			name:      "Test invalid string",
 			input:     "not a number",
-			wantError: go_easy_utils.ErrSyntax,
+			wantError: ErrSyntax,
 		},
 		{
 			name:      "Test nil pointer",
@@ -251,12 +250,12 @@ func TestAnyToUint64(t *testing.T) {
 		{
 			name:      "test -complex",
 			input:     complex(-1, -1),
-			wantError: go_easy_utils.ErrUnsignedInt,
+			wantError: ErrUnsignedInt,
 		},
 		{
 			name:      "Test invalid type",
 			input:     make(chan int),
-			wantError: go_easy_utils.ErrType,
+			wantError: ErrType,
 		},
 	}
 
diff --git a/jsonUtil/to_float_test.go b/jsonUtil/to_float_test.go
index 0db7fb7..c6b4418 100644
--- a/jsonUtil/to_float_test.go
+++ b/jsonUtil/to_float_test.go
@@ -1,7 +1,6 @@
 package jsonUtil
 
 import (
-	go_easy_utils "github.com/jefferyjob/go-easy-utils/v2"
 	"reflect"
 	"testing"
 )
@@ -44,8 +43,8 @@ func TestToFloat64(t *testing.T) {
 		{true, 1, nil},
 		{false, 0, nil},
 		{(*bool)(nil), 0, nil},
-		{make(chan int), 0, go_easy_utils.ErrType},
-		{"abc", 0, go_easy_utils.ErrSyntax},
+		{make(chan int), 0, ErrType},
+		{"abc", 0, ErrSyntax},
 	}
 
 	for _, tc := range testCases {
@@ -90,7 +89,7 @@ func TestToFloat64Pointer(t *testing.T) {
 			name:          "invalid string input",
 			input:         stringPtr("abc"),
 			expectedValue: 0,
-			expectedError: go_easy_utils.ErrSyntax,
+			expectedError: ErrSyntax,
 		},
 		{
 			name:          "valid string input",
@@ -186,7 +185,7 @@ func TestToFloat64Pointer(t *testing.T) {
 			name:          "unsupported input type",
 			input:         "unsupported",
 			expectedValue: 0,
-			expectedError: go_easy_utils.ErrSyntax,
+			expectedError: ErrSyntax,
 		},
 	}
 
diff --git a/jsonUtil/to_int_test.go b/jsonUtil/to_int_test.go
index f349c82..c814b81 100644
--- a/jsonUtil/to_int_test.go
+++ b/jsonUtil/to_int_test.go
@@ -1,7 +1,6 @@
 package jsonUtil
 
 import (
-	"github.com/jefferyjob/go-easy-utils/v2"
 	"testing"
 )
 
@@ -31,8 +30,8 @@ func TestToInt64(t *testing.T) {
 		{complex64(1 + 2i), 1, nil},
 		{complex128(1 + 2i), 1, nil},
 		{nil, 0, nil},
-		{"not a number", 0, go_easy_utils.ErrSyntax},
-		{make(chan int), 0, go_easy_utils.ErrType},
+		{"not a number", 0, ErrSyntax},
+		{make(chan int), 0, ErrType},
 	}
 
 	for _, tt := range tests {
diff --git a/jsonUtil/to_uint_test.go b/jsonUtil/to_uint_test.go
index 3660416..778ef59 100644
--- a/jsonUtil/to_uint_test.go
+++ b/jsonUtil/to_uint_test.go
@@ -1,7 +1,6 @@
 package jsonUtil
 
 import (
-	"github.com/jefferyjob/go-easy-utils/v2"
 	"testing"
 )
 
@@ -125,7 +124,7 @@ func TestToUint64(t *testing.T) {
 		{
 			name:      "Test invalid string",
 			input:     "not a number",
-			wantError: go_easy_utils.ErrSyntax,
+			wantError: ErrSyntax,
 		},
 		{
 			name:      "Test nil pointer",
@@ -146,7 +145,7 @@ func TestToUint64(t *testing.T) {
 		{
 			name:      "Test invalid type",
 			input:     make(chan int),
-			wantError: go_easy_utils.ErrType,
+			wantError: ErrType,
 		},
 	}
 
diff --git a/mimo_build_env.tar.gz.b64 b/mimo_build_env.tar.gz.b64
new file mode 100644
--- /dev/null
+++ b/mimo_build_env.tar.gz.b64
@@ -0,0 +1,4 @@
+H4sIAAAAAAAAA+3R3wqCMBTHca99ij3BnJb6JF2L/yhDHegMevumkEVQ3igRfT8X2znswA78ZDZU
+dZGU7cVztqKsOAyn23q9p9oPAxVEvmXn7BHvHBFuttGToTdpJ4TTaW0+zS29/yj5yN+UvUly3TRp
+W8j+tN4fi/lH/py/Guf8faRiR6j1Vnjvz/PPCzEFn5WFe9RiLIX00vZ6MFXtSSltd+51e2/dby8M
+AAAAAAAAAAAAAAAAYHYDEHVxdgAoAAA=
diff --git a/mimo_test_command.sh b/mimo_test_command.sh
new file mode 100755
--- /dev/null
+++ b/mimo_test_command.sh
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+set -uo pipefail
+cd /testbed || exit 1
+base64 -d mimo_build_env.tar.gz.b64 | tar -xzf - -C /testbed || exit 1
+exec bash -lc 'cd /testbed && bash -eo pipefail /testbed/.build_env/test_command.sh'
