diff --git a/robo/test/test_ei.py b/robo/test/test_ei.py
index 5b0d7ca..7eef3bc 100644
--- a/robo/test/test_ei.py
+++ b/robo/test/test_ei.py
@@ -8,6 +8,7 @@
 import GPy
 from robo.models.GPyModel import GPyModel
 from robo.acquisition.EI import EI
+from robo.recommendation.incumbent import compute_incumbent
 
 here = os.path.abspath(os.path.dirname(__file__))
 
@@ -15,46 +16,35 @@
 class EITestCase1(unittest.TestCase):
 
     def setUp(self):
-        self.x = np.array([[ 0.62971589], [ 0.63273273], [ 0.17867868], [ 0.17447447], [ 1.88558559]]);
-        self.y = np.array([[-3.69925653], [-3.66221988], [-3.65560591], [-3.58907791], [-8.06925984]]);
-        self.kernel = GPy.kern.RBF(input_dim=1, variance= 30.1646253727, lengthscale = 0.435343653946)    
-        self.noise = 1e-10
+        self.x = np.array([[0.62971589], [0.63273273], [0.17867868], [0.17447447], [1.88558559]])
+        self.y = np.array([[-3.69925653], [-3.66221988], [-3.65560591], [-3.58907791], [-8.06925984]])
+        self.kernel = GPy.kern.RBF(input_dim=1, variance=30.1646253727, lengthscale=0.435343653946)
+        self.noise = 1e-20
         self.model = GPyModel(self.kernel, noise_variance=self.noise, optimize=False)
         self.model.train(self.x, self.y)
-        self.bigger_noise = 1e2
-        self.noise_model = GPyModel(self.kernel, noise_variance=self.bigger_noise, optimize=False)
-        self.noise_model.train(self.x, self.y)
 
     def test(self):
-        X_upper = np.array([ 2.1])
+        X_upper = np.array([2.1])
         X_lower = np.array([-2.1])
-        ei_par_0 = EI(self.model, X_upper=X_upper, X_lower=X_lower,  par=0.0, derivative=True)
-        ei_par_1 = EI(self.model, X_upper=X_upper, X_lower=X_lower,  par=1.0, derivative=True)
-        ei_par_2 = EI(self.model, X_upper=X_upper, X_lower=X_lower,  par=2.0, derivative=True)
-        x_values = [0.62971589] + [2.1 * random.random() - 2.1 for i in range(10)]
-        out0 = np.array([ ei_par_0(np.array([[x]]), derivative=True) for x in x_values])
-        value0 = out0[:,0]
-        derivative0 = out0[:,1]
-        
-        out1 = np.array([ ei_par_1(np.array([[x]]), derivative=True) for x in x_values])
-        value1 = out1[:,0]
-        derivative1 = out1[:,1]
-        
-        out2 = np.array([ ei_par_2(np.array([[x]]), derivative=True) for x in x_values])
-        value2 = out2[:,0]
-        derivative2 = out2[:,1]
+
+        best = np.argmax(self.y)
+        incumbent = self.x[best]
+
+        ei_par = EI(self.model, X_upper=X_upper, X_lower=X_lower, compute_incumbent=compute_incumbent, par=0.0)
+
+        out0 = ei_par(incumbent[:, np.newaxis], derivative=True)
+        value0 = out0[0]
+        derivative0 = out0[1]
         assert(value0[0] <= 1e-5)
-        assert(np.all(value0 >= value1))
-        assert(np.all(value1 >= value2))
-        assert(np.all(np.abs(derivative0) >= np.abs(derivative1)))
-        assert(np.all(np.abs(derivative1) >= np.abs(derivative2)))
-        ei_par_0.update(self.noise_model)
-        out0_noise = np.array([ ei_par_0(np.array([[x]]), derivative=True) for x in x_values])
-        value0_noise = out0[:,0]
-        derivative0_noise = out0[:,1]
-        assert(np.all(np.abs(value0_noise) >= np.abs(value0)))
-        
-
-
-if __name__=="__main__":
+
+        x_value = incumbent + np.random.random_integers(1, 10) / 1000.
+        out1 = ei_par(x_value[:, np.newaxis], derivative=True)
+        value1 = out1[0]
+        derivative1 = out1[1]
+
+        assert(np.all(value0 < value1))
+        assert(np.all(np.abs(derivative0) < np.abs(derivative1)))
+
+
+if __name__ == "__main__":
     unittest.main()
diff --git a/robo/test/test_logei.py b/robo/test/test_logei.py
index 96fb5a3..afe166b 100644
--- a/robo/test/test_logei.py
+++ b/robo/test/test_logei.py
@@ -8,39 +8,32 @@
 import GPy
 from robo.models.GPyModel import GPyModel
 from robo.acquisition.LogEI import LogEI
-
+from robo.recommendation.incumbent import compute_incumbent
 
 here = os.path.abspath(os.path.dirname(__file__))
 
+
 class LogEITestCase1(unittest.TestCase):
+
     def setUp(self):
-        self.x = np.array([[ 0.62971589], [ 0.63273273], [ 0.17867868], [ 0.17447447], [ 1.88558559]]);
-        self.y = np.array([[-3.69925653], [-3.66221988], [-3.65560591], [-3.58907791], [-8.06925984]]);
-        self.kernel = GPy.kern.RBF(input_dim=1, variance= 30.1646253727, lengthscale = 0.435343653946)    
-        self.noise = 1e-10
+        self.x = np.array([[0.62971589], [0.63273273], [0.17867868], [0.17447447], [1.88558559]])
+        self.y = np.array([[-3.69925653], [-3.66221988], [-3.65560591], [-3.58907791], [-8.06925984]])
+        self.kernel = GPy.kern.RBF(input_dim=1, variance=30.1646253727, lengthscale=0.435343653946)
+        self.noise = 1e-20
         self.model = GPyModel(self.kernel, noise_variance=self.noise, optimize=False)
         self.model.train(self.x, self.y)
-        self.bigger_noise = 1e2
-        self.noise_model = GPyModel(self.kernel, noise_variance=self.bigger_noise, optimize=False)
-        self.noise_model.train(self.x, self.y)
 
     def test(self):
-        X_upper = np.array([ 2.1])
+        X_upper = np.array([2.1])
         X_lower = np.array([-2.1])
-        log_ei_par_0 = LogEI(self.model, X_upper=X_upper, X_lower=X_lower,  par=0.0, derivative=True)
-        log_ei_par_1 = LogEI(self.model, X_upper=X_upper, X_lower=X_lower,  par=1.0, derivative=True)
-        log_ei_par_2 = LogEI(self.model, X_upper=X_upper, X_lower=X_lower,  par=2.0, derivative=True)
-        x_values = [0.62971589] + [2.1 * random.random() - 2.1 for i in range(10)]
-        value0 = np.array([ log_ei_par_0(np.array([[x]])) for x in x_values])
-        value1 = np.array([ log_ei_par_1(np.array([[x]])) for x in x_values])
-        value2 = np.array([ log_ei_par_2(np.array([[x]])) for x in x_values])
-        
-        assert(value0[0] <= 1e-5)
-        assert(np.all(np.logical_or(value0 >= value1, np.isinf(value0), np.isinf(value1))))
-        assert(np.all(value1 >= value2))
-        log_ei_par_0.update(self.noise_model)
-        value0_noise = np.array([ log_ei_par_0(np.array([[x]])) for x in x_values])
-        assert(np.all(np.logical_or(value0_noise >= value0, np.isinf(value0), np.isinf(value0_noise))))
+
+        x_test = np.array([[1.7], [2.0]])
+        log_ei_estimator = LogEI(self.model, X_lower, X_upper, compute_incumbent=compute_incumbent)
+
+        print log_ei_estimator(x_test[0, np.newaxis])[0]
+        assert log_ei_estimator(x_test[1, np.newaxis])[0] > 0.0
+
+        self.assertAlmostEqual(log_ei_estimator(self.x[-1, np.newaxis], incumbent=np.array([1.88558559]))[0], 0.0, delta=10E-5)
 
 
 if __name__=="__main__":
diff --git a/robo/test/test_pi.py b/robo/test/test_pi.py
index 477ee78..aaba05c 100644
--- a/robo/test/test_pi.py
+++ b/robo/test/test_pi.py
@@ -10,55 +10,34 @@
 from robo.acquisition.PI import PI
 from robo.util.visualization import Visualization
 import matplotlib.pyplot as plt
-class Dummy(object):
-    pass
+
+from robo.recommendation.incumbent import compute_incumbent
+
 
 here = os.path.abspath(os.path.dirname(__file__))
 
+
 #@unittest.skip("skip first test\n")
 class PITestCase1(unittest.TestCase):
     def setUp(self):
-        self.x = np.array([[ 0.62971589], [ 0.63273273], [ 0.17867868], [ 0.17447447], [ 1.88558559]]);
-        self.y = np.array([[-3.69925653], [-3.66221988], [-3.65560591], [-3.58907791], [-8.06925984]]);
-        self.kernel = GPy.kern.RBF(input_dim=1, variance= 30.1646253727, lengthscale = 0.435343653946)    
-        self.noise = 1e-10
+        self.x = np.array([[0.62971589], [0.63273273], [0.17867868], [0.17447447], [1.88558559]])
+        self.y = np.array([[-3.69925653], [-3.66221988], [-3.65560591], [-3.58907791], [-8.06925984]])
+        self.kernel = GPy.kern.RBF(input_dim=1, variance=30.1646253727, lengthscale=0.435343653946)
+        self.noise = 1e-20
         self.model = GPyModel(self.kernel, noise_variance=self.noise, optimize=False)
         self.model.train(self.x, self.y)
-        self.bigger_noise = 1e2
-        self.noise_model = GPyModel(self.kernel, noise_variance=self.bigger_noise, optimize=False)
-        self.noise_model.train(self.x, self.y)
 
     def test(self):
-        X_upper = np.array([ 2.1])
+        X_upper = np.array([2.1])
         X_lower = np.array([-2.1])
-        pi_par_0 = PI(self.model, X_upper=X_upper, X_lower=X_lower,  par=0.0, derivative=True)
-        pi_par_1 = PI(self.model, X_upper=X_upper, X_lower=X_lower,  par=1.0, derivative=True)
-        pi_par_2 = PI(self.model, X_upper=X_upper, X_lower=X_lower,  par=2.0, derivative=True)
-        x_values = [0.62971589] + [2.1 * random.random() - 2.1 for i in range(10)]
-        out0 = np.array([ pi_par_0(np.array([[x]]), derivative=True) for x in x_values])
-        value0 = out0[:,0]
-        derivative0 = out0[:,1]
-        
-        out1 = np.array([ pi_par_1(np.array([[x]]), derivative=True) for x in x_values])
-        value1 = out1[:,0]
-        derivative1 = out1[:,1]
-        
-        out2 = np.array([ pi_par_2(np.array([[x]]), derivative=True) for x in x_values])
-        value2 = out2[:,0]
-        derivative2 = out2[:,1]
-        
-        assert(value0[0] <= 1e-5)
-        assert(np.all(value0 >= value1))
-        assert(np.all(value1 >= value2))
-        assert(np.all(np.abs(derivative0) >= np.abs(derivative1)))
-        assert(np.all(np.abs(derivative1) >= np.abs(derivative2)))
-        
-        pi_par_0.update(self.noise_model)
-        out0_noise = np.array([ pi_par_0(np.array([[x]]), derivative=True) for x in x_values])
-        value0_noise = out0[:,0]
-        derivative0_noise = out0[:,1]
-        assert(np.all(np.abs(value0_noise) >= np.abs(value0)))
 
+        x_test = np.array([[1.7], [2.0]])
+        pi_estimator = PI(self.model, X_lower, X_upper)
+
+        assert pi_estimator(x_test[0, np.newaxis], incumbent=np.array([1.88558559]))[0] > 0.0
+        assert pi_estimator(x_test[1, np.newaxis], incumbent=np.array([1.88558559]))[0] > 0.0
+
+        self.assertAlmostEqual(pi_estimator(self.x[-1, np.newaxis], incumbent=np.array([1.88558559]))[0], 0.0, delta=10E-5)
 
 if __name__=="__main__":
     unittest.main()
diff --git a/mimo_build_env.tar.gz.b64 b/mimo_build_env.tar.gz.b64
new file mode 100644
--- /dev/null
+++ b/mimo_build_env.tar.gz.b64
@@ -0,0 +1,10 @@
+H4sIAAAAAAAAA+2V32/aMBDHefZfcWJIAan5TWCqxAPbkCpt3appb1WFAjHEGrGj2GTlv6/tEgpo
+tHugq6bd5yGJ7cvd2Xf62put2SqbUl77rdci0AyTxL41x2/7HSZRGCbDOE6iVhCGgyBpQfJqGe2x
+liqtAFqVEOo5u5fW/1G8p/pXaz5VVCrplZuzxjAFHvT7p+ofBkHc1D8KYm0X9sNh3ILgrFmc4D+v
+PytKUSmQG0m2n0IS8g6+00LUFHzTEDOawaIShbHyylTlIAWonIJkirplOv+ZLqmEmlaSCQ5iAZWY
+CWAS1pJmZPfXCG5LWIgKSmD8yRlbABcKSs9UQslfTOVdpwns9O50NhMu19VxPO0/VTYPnRxXxKxO
+d6sjcHxTM193s+nubXbSj7yhF773V2zmlxuVC64n/APPDtEpHXozCe7lfElA04w8xiWtVDe4OPyr
+R5oz1YH0duxJ0Humuo9jr0gZ795aX7sN++bs7MA+ppTpDTgXzxutxPKP7Mp9I7d2yF2vR966BZE3
+ZE//bYfMRVGkPPNkfr4YL93/kV5r7v8oCYz+D6IA9f9vMM92Gk/ovRWrm/GPq1G7c/XtetKI54zx
+y46ZbxNapytod7p2QWsiU+D22uT65suH8cfPk6+fRuPlEh6VFdxiK33wO1U7nmxU7Hjeqha4NSoV
+giAIgiAIgiAIgiAIgiAIgiAIgiDIaR4A1bmBSwAoAAA=
diff --git a/mimo_test_command.sh b/mimo_test_command.sh
new file mode 100755
--- /dev/null
+++ b/mimo_test_command.sh
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+set -uo pipefail
+cd /testbed || exit 1
+base64 -d mimo_build_env.tar.gz.b64 | tar -xzf - -C /testbed || exit 1
+exec bash -lc 'cd /testbed && bash -eo pipefail /testbed/.build_env/test_command.sh'
