diff --git a/src/Sculpin/Core/Tests/Source/CompositeDataSourceTest.php b/src/Sculpin/Core/Tests/Source/CompositeDataSourceTest.php
index 9ecf884f..f26546fd 100644
--- a/src/Sculpin/Core/Tests/Source/CompositeDataSourceTest.php
+++ b/src/Sculpin/Core/Tests/Source/CompositeDataSourceTest.php
@@ -62,9 +62,9 @@ public function testDataSourceId()
 
         $dataSource = new CompositeDataSource([$ds000, $ds001, $ds002]);
 
-        $this->assertContains('TestDataSource:000', $dataSource->dataSourceId());
-        $this->assertContains('TestDataSource:001', $dataSource->dataSourceId());
-        $this->assertContains('TestDataSource:002', $dataSource->dataSourceId());
+        $this->assertStringContainsString('TestDataSource:000', $dataSource->dataSourceId());
+        $this->assertStringContainsString('TestDataSource:001', $dataSource->dataSourceId());
+        $this->assertStringContainsString('TestDataSource:002', $dataSource->dataSourceId());
     }
 
     public function testRefresh()
diff --git a/src/Sculpin/Core/Tests/Source/Map/CalculatedDateFromFilenameMapTest.php b/src/Sculpin/Core/Tests/Source/Map/CalculatedDateFromFilenameMapTest.php
index dc7b8da0..0b8bd393 100644
--- a/src/Sculpin/Core/Tests/Source/Map/CalculatedDateFromFilenameMapTest.php
+++ b/src/Sculpin/Core/Tests/Source/Map/CalculatedDateFromFilenameMapTest.php
@@ -25,7 +25,7 @@ class CalculatedDateFromFilenameMapTest extends TestCase
      */
     private $map;
 
-    protected function setUp()
+    protected function setUp(): void
     {
         $this->map = new CalculatedDateFromFilenameMap();
     }
diff --git a/src/Sculpin/Tests/Functional/FunctionalTestCase.php b/src/Sculpin/Tests/Functional/FunctionalTestCase.php
index 3935f921..90b2a35e 100644
--- a/src/Sculpin/Tests/Functional/FunctionalTestCase.php
+++ b/src/Sculpin/Tests/Functional/FunctionalTestCase.php
@@ -25,6 +25,9 @@ class FunctionalTestCase extends TestCase
     /** @var string */
     protected $executeOutput;
 
+    /** @var string */
+    protected $errorOutput;
+
     public static function setUpBeforeClass(): void
     {
         static::$fs = new Filesystem();
@@ -73,9 +76,15 @@ protected function tearDownTestProject(): void
      */
     protected function executeSculpin($command): void
     {
-        $binPath    = __DIR__ . '/../../../../bin';
-        $projectDir = static::projectDir();
-        exec("$binPath/sculpin $command --project-dir $projectDir --env=test", $this->executeOutput);
+        $process = $this->executeSculpinAsync($command, false);
+        $process->run();
+        $this->executeOutput = $process->getOutput();
+        $this->errorOutput = $process->getErrorOutput();
+
+        // This is a temporary assertion for diagnostic purposes.
+        // When adding tests that analyze error output,
+        // you are welcome to move this to its own specific test.
+        self::assertSame('', $this->errorOutput, 'Execution output should not contain errors');
     }
 
     /**
@@ -215,7 +224,7 @@ protected function assertGeneratedFileHasContent(string $filePath, string $expec
         $this->assertTrue($fileExists, $msg . ' (File Not Found!)');
 
         $contents = file_get_contents($fullPath);
-        $this->assertContains($expected, $contents, $msg);
+        $this->assertStringContainsString($expected, $contents, $msg);
     }
 
     /**
diff --git a/src/Sculpin/Tests/Functional/GenerateCommandTest.php b/src/Sculpin/Tests/Functional/GenerateCommandTest.php
index 6e7dcef5..618e4664 100644
--- a/src/Sculpin/Tests/Functional/GenerateCommandTest.php
+++ b/src/Sculpin/Tests/Functional/GenerateCommandTest.php
@@ -10,7 +10,7 @@ class GenerateCommandTest extends FunctionalTestCase
         . DIRECTORY_SEPARATOR . 'config'
         . DIRECTORY_SEPARATOR . 'sculpin_kernel.yml';
 
-    public function tearDown()
+    public function tearDown(): void
     {
         parent::tearDown();
 
diff --git a/src/Sculpin/Tests/Functional/GenerateFromMarkdownTest.php b/src/Sculpin/Tests/Functional/GenerateFromMarkdownTest.php
index eebccbf9..c59f6833 100644
--- a/src/Sculpin/Tests/Functional/GenerateFromMarkdownTest.php
+++ b/src/Sculpin/Tests/Functional/GenerateFromMarkdownTest.php
@@ -25,7 +25,7 @@ public function shouldGenerateHtmlContentFromMarkdown(): void
 
         $crawler = $this->crawlGeneratedProjectFile('/hello_world/index.html');
 
-        $this->assertContains('Hello World', $crawler->filter('h1')->text());
+        $this->assertStringContainsString('Hello World', $crawler->filter('h1')->text());
     }
 
     /** @test */
@@ -66,7 +66,7 @@ public function shouldGenerateHtmlUsingALayout()
             $pageContentEl->count(),
             "Expected generated file to have a single .page-content element."
         );
-        $this->assertContains('Hello World', $pageContentEl->text());
+        $this->assertStringContainsString('Hello World', $pageContentEl->text());
     }
 
     /** @test */
@@ -116,8 +116,8 @@ public function shouldRefreshGeneratedHtmlAfterFilesystemChange(): void
             "Expected generated file to have a single .header element."
         );
 
-        $this->assertContains($expectedHeader, $pageHeaderEl->text());
-        $this->assertContains($expectedContent, $pageContentEl->text());
+        $this->assertStringContainsString($expectedHeader, $pageHeaderEl->text());
+        $this->assertStringContainsString($expectedContent, $pageContentEl->text());
 
         // update the content
         $originalHeader  = $expectedHeader;
@@ -142,7 +142,7 @@ public function shouldRefreshGeneratedHtmlAfterFilesystemChange(): void
             "Expected generated file to have a single .page-content element."
         );
 
-        $this->assertContains($expectedContent, $pageContentEl->text());
+        $this->assertStringContainsString($expectedContent, $pageContentEl->text());
 
         // test that layouts/views refresh properly
         $this->addProjectFile($layoutFile, $layoutContent);
@@ -157,7 +157,10 @@ public function shouldRefreshGeneratedHtmlAfterFilesystemChange(): void
             "Expected generated file to have a single .header element."
         );
 
-        $this->assertContains($expectedHeader, $pageHeaderEl->text()); // I don't get it. This should be failing.
+        $this->assertStringContainsString(
+            $expectedHeader,
+            $pageHeaderEl->text()
+        );
 
         $process->stop(0);
     }
@@ -196,10 +199,13 @@ public function shouldSkipContentTypeFilesWithNoExtension(): void
 
         $this->executeSculpin('generate');
 
-        $actualOutput = implode("\n", $this->executeOutput);
-        $this->assertContains('Skipping empty or unknown file: _posts/hello_world' . PHP_EOL, $actualOutput);
-        $this->assertContains('Skipping empty or unknown file: _posts/hello_world2', $actualOutput);
-        $this->assertNotContains('Skipping empty or unknown file: _posts/hello_world3.md', $actualOutput);
+        $actualOutput = $this->executeOutput;
+        $this->assertStringContainsString(
+            'Skipping empty or unknown file: _posts/hello_world' . PHP_EOL,
+            $actualOutput
+        );
+        $this->assertStringContainsString('Skipping empty or unknown file: _posts/hello_world2', $actualOutput);
+        $this->assertStringNotContainsString('Skipping empty or unknown file: _posts/hello_world3.md', $actualOutput);
 
         $this->assertProjectLacksFile('/output_test/_posts/hello_world');
         $this->assertProjectLacksFile('/output_test/_posts/hello_world2');
@@ -230,10 +236,10 @@ public function shouldSkipHiddenFilesSilently(): void
 
         $this->executeSculpin('generate');
 
-        $actualOutput = implode("\n", $this->executeOutput);
-        $this->assertNotContains('.DS_Store', $actualOutput);
-        $this->assertNotContains('.hello_world2.swp', $actualOutput);
-        $this->assertNotContains('Skipping empty or unknown file:', $actualOutput);
+        $actualOutput = $this->executeOutput;
+        $this->assertStringNotContainsString('.DS_Store', $actualOutput);
+        $this->assertStringNotContainsString('.hello_world2.swp', $actualOutput);
+        $this->assertStringNotContainsString('Skipping empty or unknown file:', $actualOutput);
 
         $this->assertProjectLacksFile('/output_test/_posts/.DS_Store');
         $this->assertProjectLacksFile('/output_test/_posts/.hello_world2.swp');
diff --git a/mimo_build_env.tar.gz.b64 b/mimo_build_env.tar.gz.b64
new file mode 100644
--- /dev/null
+++ b/mimo_build_env.tar.gz.b64
@@ -0,0 +1,9 @@
+H4sIAAAAAAAAA+3U3UvcQBAA8HvOXzEgSPuQT81Z8PHKtX0QpPooyN7u3GUxmU13Nx6++Ld3kmux
+LaggJyKd30M+dpPZyUySbDXY1lwj3eaz11Kwk7qe9uzf/XRc1lVRzavqqD6aFWU5L8oZ1K+W0R+G
+EJUHmHnn4lPXPTf/TmUP/Y8Y4rV2XafIZKHZ3xrP9L+sipOH/h/zeHk8L7n/xf5SeNx/3n9tYGr8
+Ck2SHMA34nK0LRjskQySthhg7V0HsUFonb6BtW0RPmwbqxuwAXoVdYMGooNuPJwu1M7wplG0wfAx
+47iXf93NFd9gDHD+9RzuP2XFKWwRhoCQpnZDzmPatyqune9Sjz8CBEsap8B+oGg75KXuYIVAuEWf
+JfzS9i6gB/sr/TQll1qK6JWO1tEjgccn/j7QFHmXE4y1mJIMSd/0cMtVcD5fWcr5dCAb4SoBCF7n
+F3poex5fcNj8km8L+YUbvEYeGdOxET+rqHZj43w2BrxKDg/hxZHPVJ8vVMvzKqLh8Ljk3iw5XVId
+8uyL1tktsRxoqpVq8y9IXLmIi93PYL9Bx4zPlL8xbku/Iydv/RkIIYQQQgghhBBCCCGEEEIIIYQQ
+QrxbPwHF6pX3ACgAAA==
diff --git a/mimo_test_command.sh b/mimo_test_command.sh
new file mode 100755
--- /dev/null
+++ b/mimo_test_command.sh
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+set -uo pipefail
+cd /testbed || exit 1
+base64 -d mimo_build_env.tar.gz.b64 | tar -xzf - -C /testbed || exit 1
+exec bash -lc 'cd /testbed && bash -eo pipefail /testbed/.build_env/test_command.sh'
