diff --git a/internal/catalog/build_test.go b/internal/catalog/build_test.go
index f2df039d7f..da3bd628f0 100644
--- a/internal/catalog/build_test.go
+++ b/internal/catalog/build_test.go
@@ -45,6 +45,43 @@ func TestUpdate(t *testing.T) {
 				},
 			},
 		},
+		{
+			`
+			CREATE TYPE status AS ENUM ('open', 'closed');
+			ALTER TYPE status RENAME VALUE 'closed' TO 'shut';
+			`,
+			pg.Catalog{
+				Schemas: map[string]pg.Schema{
+					"public": {
+						Types: map[string]pg.Type{
+							"status": pg.Enum{
+								Name: "status",
+								Vals: []string{"open", "shut"},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			`
+			CREATE TYPE status AS ENUM ('open', 'closed');
+			ALTER TYPE status ADD VALUE 'unknown';
+			ALTER TYPE status ADD VALUE IF NOT EXISTS 'unknown';
+			`,
+			pg.Catalog{
+				Schemas: map[string]pg.Schema{
+					"public": {
+						Types: map[string]pg.Type{
+							"status": pg.Enum{
+								Name: "status",
+								Vals: []string{"open", "closed", "unknown"},
+							},
+						},
+					},
+				},
+			},
+		},
 		{
 			"CREATE TABLE venues ();",
 			pg.Catalog{
@@ -240,6 +277,27 @@ func TestUpdate(t *testing.T) {
 				},
 			},
 		},
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
 		{
 			`
 			CREATE TABLE venues ();
diff --git a/internal/postgresql/catalog_test.go b/internal/postgresql/catalog_test.go
index 61619794fa..abc1a26477 100644
--- a/internal/postgresql/catalog_test.go
+++ b/internal/postgresql/catalog_test.go
@@ -17,11 +17,13 @@ import (
 func TestUpdate(t *testing.T) {
 	p := NewParser()
 
-	for i, tc := range []struct {
+	for _, tc := range []struct {
+		name string
 		stmt string
 		s    *catalog.Schema
 	}{
 		{
+			"create-enum",
 			"CREATE TYPE status AS ENUM ('open', 'closed');",
 			&catalog.Schema{
 				Name: "public",
@@ -34,6 +36,40 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"alter-type-rename-value",
+			`
+			CREATE TYPE status AS ENUM ('open', 'closed');
+			ALTER TYPE status RENAME VALUE 'closed' TO 'shut';
+			`,
+			&catalog.Schema{
+				Name: "public",
+				Types: []catalog.Type{
+					&catalog.Enum{
+						Name: "status",
+						Vals: []string{"open", "shut"},
+					},
+				},
+			},
+		},
+		{
+			"alter-type-add-value",
+			`
+			CREATE TYPE status AS ENUM ('open', 'closed');
+			ALTER TYPE status ADD VALUE 'unknown';
+			ALTER TYPE status ADD VALUE IF NOT EXISTS 'unknown';
+			`,
+			&catalog.Schema{
+				Name: "public",
+				Types: []catalog.Type{
+					&catalog.Enum{
+						Name: "status",
+						Vals: []string{"open", "closed", "unknown"},
+					},
+				},
+			},
+		},
+		{
+			"create-table",
 			"CREATE TABLE venues ();",
 			&catalog.Schema{
 				Name: "public",
@@ -45,6 +81,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"alter-table-drop-column",
 			`
 			CREATE TABLE foo ();
 			ALTER TABLE foo ADD COLUMN bar text;
@@ -60,6 +97,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"alter-table-drop-column-if-exists",
 			`
 			CREATE TABLE foo ();
 			ALTER TABLE foo DROP COLUMN IF EXISTS bar;
@@ -74,6 +112,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"alter-table-set-not-null",
 			`
 			CREATE TABLE foo (bar text);
 			ALTER TABLE foo ALTER bar SET NOT NULL;
@@ -116,6 +155,7 @@ func TestUpdate(t *testing.T) {
 			},
 		*/
 		{
+			"alter-table-drop-not-null",
 			`
 			CREATE TABLE foo (bar text NOT NULL);
 			ALTER TABLE foo ALTER bar DROP NOT NULL;
@@ -136,6 +176,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"alter-table-column-drop-not-null",
 			`
 			CREATE TABLE foo (bar text NOT NULL);
 			ALTER TABLE foo ALTER COLUMN bar DROP NOT NULL;
@@ -156,6 +197,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"alter-table-rename-column",
 			`
 			CREATE TABLE foo (bar text);
 			ALTER TABLE foo RENAME bar TO baz;
@@ -176,6 +218,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"alter-table-set-data-type",
 			`
 			CREATE TABLE foo (bar text);
 			ALTER TABLE foo ALTER bar SET DATA TYPE bool;
@@ -196,6 +239,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"alter-table-set-schema",
 			`
 			CREATE SCHEMA foo;
 			CREATE TABLE bar ();
@@ -211,6 +255,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"drop-table",
 			`
 			CREATE TABLE venues ();
 			DROP TABLE venues;
@@ -218,6 +263,7 @@ func TestUpdate(t *testing.T) {
 			nil,
 		},
 		{
+			"drop-table-if-exists",
 			`
 			CREATE TABLE venues ();
 			DROP TABLE IF EXISTS venues;
@@ -226,6 +272,7 @@ func TestUpdate(t *testing.T) {
 			nil,
 		},
 		{
+			"alter-table-rename",
 			`
 			CREATE TABLE venues ();
 			ALTER TABLE venues RENAME TO arenas;
@@ -240,6 +287,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"drop-type",
 			`
 			CREATE TYPE status AS ENUM ('open', 'closed');
 			DROP TYPE status;
@@ -247,6 +295,7 @@ func TestUpdate(t *testing.T) {
 			nil,
 		},
 		{
+			"drop-type-if-exists",
 			`
 			CREATE TYPE status AS ENUM ('open', 'closed');
 			DROP TYPE IF EXISTS status;
@@ -255,6 +304,7 @@ func TestUpdate(t *testing.T) {
 			nil,
 		},
 		{
+			"drop-table-in-schema",
 			`
 			CREATE TABLE venues ();
 			DROP TABLE public.venues;
@@ -262,6 +312,7 @@ func TestUpdate(t *testing.T) {
 			nil,
 		},
 		{
+			"drop-type-in-schema",
 			`
 			CREATE TYPE status AS ENUM ('open', 'closed');
 			DROP TYPE public.status;
@@ -269,6 +320,7 @@ func TestUpdate(t *testing.T) {
 			nil,
 		},
 		{
+			"drop-schema",
 			`
 			CREATE SCHEMA foo;
 			DROP SCHEMA foo;
@@ -276,12 +328,14 @@ func TestUpdate(t *testing.T) {
 			nil,
 		},
 		{
+			"drop-schema-if-exists",
 			`
 			DROP SCHEMA IF EXISTS foo;
 			`,
 			nil,
 		},
 		{
+			"drop-function-if-exists",
 			`
 			DROP FUNCTION IF EXISTS bar;
 			DROP FUNCTION IF EXISTS bar();
@@ -289,6 +343,7 @@ func TestUpdate(t *testing.T) {
 			nil,
 		},
 		{
+			"alter-table-drop-constraint",
 			`
 			CREATE TABLE venues (id SERIAL PRIMARY KEY);
 			ALTER TABLE venues DROP CONSTRAINT venues_id_pkey;
@@ -310,6 +365,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"create-function",
 			`
 			CREATE FUNCTION foo(TEXT) RETURNS bool AS $$ SELECT true $$ LANGUAGE sql;
 			`,
@@ -329,6 +385,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"create-function-args",
 			`
 			CREATE FUNCTION foo(bar TEXT) RETURNS bool AS $$ SELECT true $$ LANGUAGE sql;
 			CREATE FUNCTION foo(bar TEXT, baz TEXT) RETURNS TEXT AS $$ SELECT "baz" $$ LANGUAGE sql;
@@ -364,6 +421,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"create-function-types",
 			`
 			CREATE FUNCTION foo(bar TEXT) RETURNS bool AS $$ SELECT true $$ LANGUAGE sql;
 			CREATE FUNCTION foo(bar INTEGER) RETURNS TEXT AS $$ SELECT "baz" $$ LANGUAGE sql;
@@ -395,6 +453,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"create-function-return",
 			`
 			CREATE FUNCTION foo(bar TEXT, baz TEXT="baz") RETURNS bool AS $$ SELECT true $$ LANGUAGE sql;
 			`,
@@ -420,6 +479,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"drop-function-args",
 			`
 			CREATE FUNCTION foo(bar text) RETURNS bool AS $$ SELECT true $$ LANGUAGE sql;
 			DROP FUNCTION foo(text);
@@ -427,6 +487,7 @@ func TestUpdate(t *testing.T) {
 			nil,
 		},
 		{
+			"drop-function",
 			`
 			CREATE FUNCTION foo(bar text) RETURNS bool AS $$ SELECT true $$ LANGUAGE sql;
 			DROP FUNCTION foo;
@@ -437,6 +498,7 @@ func TestUpdate(t *testing.T) {
 		// CREATE FUNCTION foo() RETURNS bool AS $$ SELECT true $$ LANGUAGE sql;
 		// DROP FUNCTION foo -- FAIL
 		{
+			"pg_temp",
 			`
 			CREATE TABLE pg_temp.migrate (val SERIAL);
 			INSERT INTO pg_temp.migrate (val) SELECT val FROM old;
@@ -458,6 +520,7 @@ func TestUpdate(t *testing.T) {
 			},
 		},
 		{
+			"comment",
 			`
 			CREATE SCHEMA foo;
 			CREATE TABLE foo.bar (baz text);
@@ -494,7 +557,7 @@ func TestUpdate(t *testing.T) {
 		},
 	} {
 		test := tc
-		t.Run(strconv.Itoa(i), func(t *testing.T) {
+		t.Run(test.name, func(t *testing.T) {
 			stmts, err := p.Parse(strings.NewReader(test.stmt))
 			if err != nil {
 				t.Log(test.stmt)
diff --git a/mimo_build_env.tar.gz.b64 b/mimo_build_env.tar.gz.b64
new file mode 100644
--- /dev/null
+++ b/mimo_build_env.tar.gz.b64
@@ -0,0 +1,5 @@
+H4sIAAAAAAAAA+3S326DIBQGcK99itPsWhEserW9SoNA1IRCJ+hev85k2Z9s602bpdn3uzkETsIJ
+H2U3j84crF9YdivVqpVyq6uvdVtzKbgQdc2bOqs4l22bkbzZRB/MMamJKJtCSL/1XTq/U+V7/snG
+dNDheFTelHG43h2X8hecv+Uv2n2z5r9veJ1Rdb0RfvbP83/YsW70rFNxyLWh7RN01uTGLkV8GZMe
+qNeaeE3iia2bzM/O5X2g10Yq2eiTnbxyTKukXOgZFQsVOsw+PfLv+k4hpn6y8dl9bv3rhwAAAAAA
+AAAAAAAAAAAAAAAAuENnr431LQAoAAA=
diff --git a/mimo_test_command.sh b/mimo_test_command.sh
new file mode 100755
--- /dev/null
+++ b/mimo_test_command.sh
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+set -uo pipefail
+cd /testbed || exit 1
+base64 -d mimo_build_env.tar.gz.b64 | tar -xzf - -C /testbed || exit 1
+exec bash -lc 'cd /testbed && bash -eo pipefail /testbed/.build_env/test_command.sh'
