diff --git a/tests/run/cpp_stl_associated_containers_contains_cpp20.pyx b/tests/run/cpp_stl_associated_containers_contains_cpp20.pyx
new file mode 100644
index 00000000000..ebe8d8fa883
--- /dev/null
+++ b/tests/run/cpp_stl_associated_containers_contains_cpp20.pyx
@@ -0,0 +1,106 @@
+# mode: run
+# tag: cpp, cpp20
+
+# cython: language_level=3
+
+from libcpp.map cimport map, multimap
+from libcpp.set cimport set, multiset
+from libcpp.unordered_map cimport unordered_map, unordered_multimap
+from libcpp.unordered_set cimport unordered_set, unordered_multiset
+
+def test_map_contains(vals, int key_to_find):
+    """
+    >>> test_map_contains([(1,100),(2,200),(3,300)], 3)
+    True
+    >>> test_map_contains([(1,100),(2,200),(3,300)], 4)
+    False
+    """
+    cdef map[int,int] m = map[int, int]()
+    for v in vals:
+        m.insert(v)
+    return m.contains(key_to_find)
+
+def test_unordered_map_contains(vals, int key_to_find):
+    """
+    >>> test_unordered_map_contains([(1,100),(2,200),(3,300)], 3)
+    True
+    >>> test_unordered_map_contains([(1,100),(2,200),(3,300)], 4)
+    False
+    """
+    cdef unordered_map[int,int] um = unordered_map[int, int]()
+    for v in vals:
+        um.insert(v)
+    return um.contains(key_to_find)
+
+def test_multimap_contains(vals, int key_to_find):
+    """
+    >>> test_multimap_contains([(1,100),(2,200),(3,300)], 3)
+    True
+    >>> test_multimap_contains([(1,100),(2,200),(3,300)], 4)
+    False
+    """
+    cdef multimap[int,int] mm = multimap[int, int]()
+    for v in vals:
+        mm.insert(v)
+    return mm.contains(key_to_find)
+
+def test_unordered_multimap_contains(vals, int key_to_find):
+    """
+    >>> test_unordered_multimap_contains([(1,100),(2,200),(3,300)], 3)
+    True
+    >>> test_unordered_multimap_contains([(1,100),(2,200),(3,300)], 4)
+    False
+    """
+    cdef unordered_multimap[int,int] umm = unordered_multimap[int, int]()
+    for v in vals:
+        umm.insert(v)
+    return umm.contains(key_to_find)
+
+
+def test_set_contains(vals, int val_to_find):
+    """
+    >>> test_set_contains([1, 2, 3], 3)
+    True
+    >>> test_set_contains([1, 2, 3], 4)
+    False
+    """
+    cdef set[int] s = set[int]()
+    for v in vals:
+        s.insert(v)
+    return s.contains(val_to_find)
+
+def test_unordered_set_contains(vals, int val_to_find):
+    """
+    >>> test_unordered_set_contains([1, 2, 3], 3)
+    True
+    >>> test_unordered_set_contains([1, 2, 3], 4)
+    False
+    """
+    cdef unordered_set[int] us = unordered_set[int]()
+    for v in vals:
+        us.insert(v)
+    return us.contains(val_to_find)
+
+def test_multiset_contains(vals, int val_to_find):
+    """
+    >>> test_multiset_contains([1, 2, 3], 3)
+    True
+    >>> test_multiset_contains([1, 2, 3], 4)
+    False
+    """
+    cdef multiset[int] ms = multiset[int]()
+    for v in vals:
+        ms.insert(v)
+    return ms.contains(val_to_find)
+
+def test_unordered_multiset_contains(vals, int val_to_find):
+    """
+    >>> test_unordered_multiset_contains([1, 2, 3], 3)
+    True
+    >>> test_unordered_multiset_contains([1, 2, 3], 4)
+    False
+    """
+    cdef unordered_multiset[int] ums = unordered_multiset[int]()
+    for v in vals:
+        ums.insert(v)
+    return ums.contains(val_to_find)
diff --git a/tests/run/cpp_stl_numeric_ops_cpp17.pyx b/tests/run/cpp_stl_numeric_ops_cpp17.pyx
index eba4d2beb86..d0e3f8b7b2 100644
--- a/tests/run/cpp_stl_numeric_ops_cpp17.pyx
+++ b/tests/run/cpp_stl_numeric_ops_cpp17.pyx
@@ -3,7 +3,7 @@
 
 from libcpp.numeric cimport (reduce, transform_reduce, inclusive_scan, 
                              exclusive_scan, transform_inclusive_scan, 
-                             transform_exclusive_scan)
+                             transform_exclusive_scan, gcd, lcm)
 from libcpp.execution cimport seq
 from libcpp.vector cimport vector
 
@@ -275,3 +275,23 @@ def test_transform_exclusive_scan_with_execpolicy(vector[int] v, int init):
     cdef vector[int] out = vector[int](v.size())
     transform_exclusive_scan(seq, v.begin(), v.end(), out.begin(), init, add_integers, multiply_with_2)
     return out
+
+
+def test_gcd(int a, int b):
+    """
+    >>> test_gcd(12, 18)
+    6
+    >>> test_gcd(21, 14)
+    7
+    """
+    return gcd[int](a, b)
+
+
+def test_lcm(int a, int b):
+    """
+    >>> test_lcm(45, 75)
+    225
+    >>> test_lcm(4, 6)
+    12
+    """
+    return lcm[int](a, b)
diff --git a/tests/run/cpp_stl_numeric_ops_cpp20.pyx b/tests/run/cpp_stl_numeric_ops_cpp20.pyx
new file mode 100644
index 00000000000..9a8f5ec4ac
--- /dev/null
+++ b/tests/run/cpp_stl_numeric_ops_cpp20.pyx
@@ -0,0 +1,26 @@
+# mode: run
+# tag: cpp, werror, cpp20
+
+from libcpp.numeric cimport midpoint
+
+
+def test_midpoint_integer(int a, int b):
+    """
+    >>> test_midpoint_integer(2, 6)
+    4
+    >>> test_midpoint_integer(10, 20)
+    15
+    """
+    cdef int res = midpoint[int](a, b)
+    return res
+
+
+def test_midpoint_float(float a, float b):
+    """
+    >>> test_midpoint_float(2, 6)
+    4.0
+    >>> test_midpoint_float(1, 2)
+    1.5
+    """
+    cdef float res = midpoint[float](a, b)
+    return res
diff --git a/tests/run/cpp_stl_string_cpp20.pyx b/tests/run/cpp_stl_string_cpp20.pyx
new file mode 100644
index 00000000000..f3a2b80d15f
--- /dev/null
+++ b/tests/run/cpp_stl_string_cpp20.pyx
@@ -0,0 +1,61 @@
+# mode: run
+# tag: cpp, werror, cpp20
+
+from libcpp cimport bool
+from libcpp.string cimport string
+
+b_A = b'A'
+b_F = b'F'
+b_abc = b"ABC"
+b_def = b"DEF"
+
+def test_string_starts_with_char(bytes py_str):
+    """
+    Test std::string.starts_with() with char type argument
+    >>> test_string_starts_with_char(b'A')
+    True
+    >>> test_string_starts_with_char(b'F')
+    False
+    """
+    cdef char c = py_str[0]
+    cdef string s = b"ABCDEF"
+    return s.starts_with(c)
+
+
+def test_string_starts_with_cstr(bytes py_str):
+    """
+    Test std::string.starts_with() with c str type argument (char*)
+    >>> test_string_starts_with_cstr(b"ABC")
+    True
+    >>> test_string_starts_with_cstr(b"DEF")
+    False
+    """
+    cdef char* c = py_str
+    cdef string s = b"ABCDEF"
+    return s.starts_with(c)
+
+
+def test_string_ends_with_char(bytes py_str):
+    """
+    Test std::string.ends_with() with char type argument
+    >>> test_string_ends_with_char(b'F')
+    True
+    >>> test_string_ends_with_char(b'A')
+    False
+    """
+    cdef char c = py_str[0]
+    cdef string s = b"ABCDEF"
+    return s.ends_with(c)
+
+
+def test_string_ends_with_cstr(bytes py_str):
+    """
+    Test std::string.ends_with() with c str type argument (char*)
+    >>> test_string_ends_with_cstr(b"DEF")
+    True
+    >>> test_string_ends_with_cstr(b"ABC")
+    False
+    """
+    cdef char* c = py_str
+    cdef string s = b"ABCDEF"
+    return s.ends_with(c)
\ No newline at end of file
diff --git a/tests/run/cpp_stl_string_cpp23.pyx b/tests/run/cpp_stl_string_cpp23.pyx
new file mode 100644
index 00000000000..74866c8a16
--- /dev/null
+++ b/tests/run/cpp_stl_string_cpp23.pyx
@@ -0,0 +1,29 @@
+# mode: run
+# tag: cpp, werror, cpp23
+# distutils: extra_compile_args=-std=c++23
+
+from libcpp.string cimport string
+
+
+def test_string_contains_char(bytes py_str):
+    """
+    >>> test_string_contains_char(b"C")
+    True
+    >>> test_string_contains_char(b"Z")
+    False
+    """
+    cdef char c = py_str[0]
+    cdef string s = b"ABCDEF"
+    return s.contains(c)
+
+
+def test_string_contains_cstr(bytes py_str):
+    """
+    >>> test_string_contains_cstr(b"BCD")
+    True
+    >>> test_string_contains_cstr(b"XYZ")
+    False
+    """
+    cdef char* c = py_str
+    cdef string s = b"ABCDEF"
+    return s.contains(c)
diff --git a/mimo_test_command.sh b/mimo_test_command.sh
new file mode 100755
--- /dev/null
+++ b/mimo_test_command.sh
@@ -0,0 +1,56 @@
+#!/bin/bash
+set -uo pipefail
+cd /testbed
+
+
+
+dev-switch python3 3.11
+
+
+if ! command -v g++-15 >/dev/null 2>&1; then
+    apt-get update
+    apt-get install -y gcc-15 g++-15
+fi
+export CC=gcc-15
+export CXX=g++-15
+
+
+if [ ! -d /testbed/.build_env/.venv ]; then
+    python3 -m venv /testbed/.build_env/.venv
+fi
+set +u
+source /testbed/.build_env/.venv/bin/activate
+set -u
+
+
+pip install --quiet setuptools
+
+
+python3 setup.py build_ext --inplace 2>&1 | tail -5
+
+
+case "${1:-}" in
+  base)
+    rm -rf TEST_TMP/run/cpp/cpp_stl_vector*
+    python3 runtests.py --no-c \
+      cpp_stl_vector
+    ;;
+  new)
+    rm -rf \
+      TEST_TMP/run/cpp/cpp_stl_numeric_ops_cpp17* \
+      TEST_TMP/run/cpp/cpp_stl_numeric_ops_cpp20* \
+      TEST_TMP/run/cpp/cpp_stl_associated_containers_contains_cpp20* \
+      TEST_TMP/run/cpp/cpp_stl_string_cpp20* \
+      TEST_TMP/run/cpp/cpp_stl_string_cpp23*
+    python3 runtests.py --no-c \
+      cpp_stl_numeric_ops_cpp17 \
+      cpp_stl_numeric_ops_cpp20 \
+      cpp_stl_associated_containers_contains_cpp20 \
+      cpp_stl_string_cpp20 \
+      cpp_stl_string_cpp23
+    ;;
+  *)
+    echo "usage: $0 base|new"
+    exit 2
+    ;;
+esac
