diff --git a/hazelcast-client/src/test/java/com/hazelcast/client/cache/ClientCacheCreationTest.java b/hazelcast-client/src/test/java/com/hazelcast/client/cache/ClientCacheCreationTest.java
new file mode 100644
index 0000000000000..6ef4069a15077
--- /dev/null
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/cache/ClientCacheCreationTest.java
@@ -0,0 +1,54 @@
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+package com.hazelcast.client.cache;
+
+import com.hazelcast.cache.CacheCreationTest;
+import com.hazelcast.client.HazelcastClient;
+import com.hazelcast.client.cache.impl.HazelcastClientCachingProvider;
+import com.hazelcast.client.config.ClientConfig;
+import com.hazelcast.config.Config;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import org.junit.Ignore;
+import org.junit.Test;
+
+import javax.cache.spi.CachingProvider;
+import java.util.Arrays;
+
+public class ClientCacheCreationTest extends CacheCreationTest {
+
+
+    @Test
+    @Ignore("Only applicable for member-side HazelcastInstance")
+    @Override
+    public void createInvalidCache_fromDeclarativeConfig_throwsException_fromHazelcastInstanceCreation() {
+
+    }
+
+    @Override
+    protected CachingProvider createCachingProvider(Config hzConfig) {
+        HazelcastInstance hazelcastInstance = Hazelcast.newHazelcastInstance(hzConfig);
+        ClientConfig clientConfig = null;
+        if (hzConfig != null) {
+            clientConfig = new ClientConfig();
+            clientConfig.getGroupConfig().setName(hzConfig.getGroupConfig().getName());
+            clientConfig.getGroupConfig().setPassword(hzConfig.getGroupConfig().getPassword());
+            clientConfig.getNetworkConfig().setAddresses(Arrays.asList("127.0.0.1"));
+        }
+        return HazelcastClientCachingProvider.createCachingProvider(HazelcastClient.newHazelcastClient(clientConfig));
+    }
+}
diff --git a/hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java b/hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java
index 8a8e68764770b..dc6b3967e1a2d 100644
--- a/hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java
@@ -17,7 +17,12 @@
 package com.hazelcast.cache;
 
 import com.hazelcast.cache.impl.HazelcastServerCachingProvider;
+import com.hazelcast.config.CacheConfig;
+import com.hazelcast.config.CacheSimpleConfig;
 import com.hazelcast.config.Config;
+import com.hazelcast.config.EvictionConfig;
+import com.hazelcast.config.EvictionPolicy;
+import com.hazelcast.config.InMemoryFormat;
 import com.hazelcast.config.XmlConfigBuilder;
 import com.hazelcast.core.Hazelcast;
 import com.hazelcast.core.HazelcastInstance;
@@ -28,21 +33,33 @@
 import org.junit.After;
 import org.junit.Before;
 import org.junit.BeforeClass;
+import org.junit.Rule;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
+import org.junit.rules.ExpectedException;
 import org.junit.runner.RunWith;
 
 import javax.cache.Cache;
+import javax.cache.CacheException;
+import javax.cache.CacheManager;
+import javax.cache.Caching;
+import javax.cache.spi.CachingProvider;
 import java.net.URISyntaxException;
 import java.net.URL;
+import java.util.Arrays;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 
+import static com.hazelcast.config.EvictionConfig.MaxSizePolicy.ENTRY_COUNT;
+
 @RunWith(HazelcastSerialClassRunner.class)
 @Category(SlowTest.class)
 public class CacheCreationTest {
 
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
     private static final int THREAD_COUNT = 4;
 
     private static Config hzConfig;
@@ -61,14 +78,14 @@ public void killAllHazelcastInstances() throws Exception {
     }
 
     @Test
-    public void creatingASingleCache() throws URISyntaxException {
-        HazelcastServerCachingProvider cachingProvider = createCachingProvider(hzConfig);
+    public void createSingleCache() throws URISyntaxException {
+        CachingProvider cachingProvider = createCachingProvider(hzConfig);
         Cache<Object, Object> cache = cachingProvider.getCacheManager().getCache("xmlCache" + 1);
         cache.get(1);
     }
 
     @Test
-    public void creatingASingleCacheFromMultiProviders() throws URISyntaxException, InterruptedException {
+    public void createOrGetConcurrentlySingleCache_fromMultiProviders() throws URISyntaxException, InterruptedException {
         ExecutorService executorService = Executors.newFixedThreadPool(THREAD_COUNT);
 
         final CountDownLatch latch = new CountDownLatch(THREAD_COUNT);
@@ -76,7 +93,7 @@ public void creatingASingleCacheFromMultiProviders() throws URISyntaxException,
             executorService.execute(new Runnable() {
                 @Override
                 public void run() {
-                    HazelcastServerCachingProvider cachingProvider = createCachingProvider(hzConfig);
+                    CachingProvider cachingProvider = createCachingProvider(hzConfig);
                     Cache<Object, Object> cache = cachingProvider.getCacheManager().getCache("xmlCache");
                     cache.get(1);
                     latch.countDown();
@@ -88,7 +105,7 @@ public void run() {
     }
 
     @Test
-    public void creatingMultiCacheFromMultiProviders() throws Exception {
+    public void createConcurrentlyMultipleCaches_fromMultipleProviders() throws Exception {
         ExecutorService executorService = Executors.newFixedThreadPool(THREAD_COUNT);
 
         final CountDownLatch latch = new CountDownLatch(THREAD_COUNT);
@@ -97,7 +114,7 @@ public void creatingMultiCacheFromMultiProviders() throws Exception {
             executorService.execute(new Runnable() {
                 @Override
                 public void run() {
-                    HazelcastServerCachingProvider cachingProvider = createCachingProvider(hzConfig);
+                    CachingProvider cachingProvider = createCachingProvider(hzConfig);
                     Cache<Object, Object> cache = cachingProvider.getCacheManager().getCache(cacheName);
                     cache.get(1);
                     latch.countDown();
@@ -108,8 +125,58 @@ public void run() {
         executorService.shutdown();
     }
 
-    private HazelcastServerCachingProvider createCachingProvider(Config hzConfig) {
+    @Test
+    public void createInvalidCache_fromProgrammaticConfig_throwsException_fromCacheManager_getCache() {
+        Config config = createInvalidConfig();
+        CachingProvider cachingProvider = createCachingProvider(config);
+        CacheManager defaultCacheManager = cachingProvider.getCacheManager();
+        thrown.expect(IllegalArgumentException.class);
+        Cache cache = defaultCacheManager.getCache("test");
+    }
+
+    @Test
+    public void createInvalidCache_throwsException_fromCacheManager_createCache() {
+        CachingProvider cachingProvider = createCachingProvider(createBasicConfig());
+        CacheManager defaultCacheManager = cachingProvider.getCacheManager();
+        thrown.expect(IllegalArgumentException.class);
+        Cache cache = defaultCacheManager.createCache("test", createInvalidCacheConfig());
+    }
+
+    @Test
+    public void createInvalidCache_fromDeclarativeConfig_throwsException_fromHazelcastInstanceCreation() {
+        System.setProperty("hazelcast.config", "classpath:test-hazelcast-invalid-cache.xml");
+        CachingProvider cachingProvider = Caching.getCachingProvider();
+        thrown.expect(CacheException.class);
+        CacheManager defaultCacheManager = cachingProvider.getCacheManager();
+    }
+
+    protected CachingProvider createCachingProvider(Config hzConfig) {
         HazelcastInstance hazelcastInstance = Hazelcast.newHazelcastInstance(hzConfig);
         return HazelcastServerCachingProvider.createCachingProvider(hazelcastInstance);
     }
+
+
+    private Config createInvalidConfig() {
+        Config config = createBasicConfig();
+        CacheSimpleConfig cacheSimpleConfig = new CacheSimpleConfig();
+        cacheSimpleConfig.setName("test");
+        cacheSimpleConfig.setInMemoryFormat(InMemoryFormat.NATIVE);
+        cacheSimpleConfig.setEvictionConfig(new EvictionConfig(1000, ENTRY_COUNT, EvictionPolicy.LFU));
+        config.addCacheConfig(cacheSimpleConfig);
+        return config;
+    }
+
+    private CacheConfig createInvalidCacheConfig() {
+        CacheConfig cacheConfig = new CacheConfig("test");
+        cacheConfig.setInMemoryFormat(InMemoryFormat.NATIVE);
+        cacheConfig.setEvictionConfig(new EvictionConfig(1000, ENTRY_COUNT, EvictionPolicy.LFU));
+        return cacheConfig;
+    }
+
+    private Config createBasicConfig() {
+        Config config = new Config();
+        config.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);
+        config.getNetworkConfig().getJoin().getTcpIpConfig().setEnabled(true).setMembers(Arrays.asList("127.0.0.1"));
+        return config;
+    }
 }
diff --git a/hazelcast/src/test/java/com/hazelcast/config/XMLConfigBuilderTest.java b/hazelcast/src/test/java/com/hazelcast/config/XMLConfigBuilderTest.java
index 8b59ad9fba21b..cb976bbbea52a 100644
--- a/hazelcast/src/test/java/com/hazelcast/config/XMLConfigBuilderTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/config/XMLConfigBuilderTest.java
@@ -1771,6 +1771,18 @@ public void testConfigPermission() {
         assertPermissionConfig(expected, config);
     }
 
+    @Test(expected = InvalidConfigurationException.class)
+    public void testCacheConfig_withInvalidEvictionConfig_failsFast() {
+        String xml = HAZELCAST_START_TAG
+                + "    <cache name=\"cache\">"
+                + "        <eviction size=\"10000000\" max-size-policy=\"ENTRY_COUNT\" eviction-policy=\"LFU\"/>"
+                + "        <in-memory-format>NATIVE</in-memory-format>\n"
+                + "    </cache>"
+                + HAZELCAST_END_TAG;
+
+        Config config = buildConfig(xml);
+    }
+
     private void assertPermissionConfig(PermissionConfig expected, Config config) {
         Iterator<PermissionConfig> permConfigs = config.getSecurityConfig().getClientPermissionConfigs().iterator();
         PermissionConfig configured = permConfigs.next();
diff --git a/hazelcast/src/test/java/com/hazelcast/internal/config/ConfigValidatorTest.java b/hazelcast/src/test/java/com/hazelcast/internal/config/ConfigValidatorTest.java
index 668c019a65d8f..eb50d3eca0f38 100644
--- a/hazelcast/src/test/java/com/hazelcast/internal/config/ConfigValidatorTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/config/ConfigValidatorTest.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.internal.config;
 
+import com.hazelcast.config.CacheSimpleConfig;
 import com.hazelcast.config.EvictionConfig;
 import com.hazelcast.config.EvictionPolicy;
 import com.hazelcast.config.InMemoryFormat;
@@ -37,6 +38,7 @@
 import static com.hazelcast.config.InMemoryFormat.OBJECT;
 import static com.hazelcast.config.NearCacheConfig.LocalUpdatePolicy.CACHE_ON_UPDATE;
 import static com.hazelcast.config.NearCacheConfig.LocalUpdatePolicy.INVALIDATE;
+import static com.hazelcast.internal.config.ConfigValidator.checkCacheConfig;
 import static com.hazelcast.internal.config.ConfigValidator.checkEvictionConfig;
 import static com.hazelcast.internal.config.ConfigValidator.checkMapConfig;
 import static com.hazelcast.internal.config.ConfigValidator.checkNearCacheConfig;
@@ -259,6 +261,22 @@ public void checkEvictionConfig_whenNoneOfTheComparatorAndComparatorClassNameAre
         checkEvictionConfig(null, null, null, true);
     }
 
+    @Test(expected = IllegalArgumentException.class)
+    public void checkCacheConfig_whenNATIVEAndEntryCountMaxSizePolicy() {
+        CacheSimpleConfig cacheSimpleConfig = new CacheSimpleConfig();
+        cacheSimpleConfig.getEvictionConfig().setMaximumSizePolicy(EvictionConfig.MaxSizePolicy.ENTRY_COUNT);
+        cacheSimpleConfig.setInMemoryFormat(NATIVE);
+        checkCacheConfig(cacheSimpleConfig);
+    }
+
+    @Test
+    public void checkCacheConfig_whenOBJECTAndEntryCountMaxSizePolicy() {
+        CacheSimpleConfig cacheSimpleConfig = new CacheSimpleConfig();
+        cacheSimpleConfig.getEvictionConfig().setMaximumSizePolicy(EvictionConfig.MaxSizePolicy.ENTRY_COUNT);
+        cacheSimpleConfig.setInMemoryFormat(OBJECT);
+        checkCacheConfig(cacheSimpleConfig);
+    }
+
     @Test
     public void test_checkNearCacheConfig_withPreLoaderConfig_onClients() {
         NearCacheConfig nearCacheConfig = getNearCacheConfig(BINARY)
diff --git a/hazelcast/src/test/resources/test-hazelcast-invalid-cache.xml b/hazelcast/src/test/resources/test-hazelcast-invalid-cache.xml
new file mode 100644
index 0000000000000..5452cee91ccd9
--- /dev/null
+++ b/hazelcast/src/test/resources/test-hazelcast-invalid-cache.xml
@@ -0,0 +1,40 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+  ~
+  ~ Licensed under the Apache License, Version 2.0 (the "License");
+  ~ you may not use this file except in compliance with the License.
+  ~ You may obtain a copy of the License at
+  ~
+  ~ http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<hazelcast xsi:schemaLocation="http://www.hazelcast.com/schema/config hazelcast-config-3.9.xsd"
+           xmlns="http://www.hazelcast.com/schema/config"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+
+    <group>
+        <name>F4D03AC8-B5F7-4542-9BAD-0A94BA48E95D</name>
+        <password>6E282022-3BAD-4F5F-8122-CCBE85BD057E</password>
+    </group>
+    <network>
+        <join>
+            <multicast enabled="false"/>
+            <tcp-ip enabled="true">
+                <member>127.0.0.1</member>
+            </tcp-ip>
+        </join>
+    </network>
+
+    <cache name="invalid-cache">
+        <in-memory-format>NATIVE</in-memory-format>
+        <eviction size="99" max-size-policy="ENTRY_COUNT" eviction-policy="LFU"/>
+    </cache>
+
+</hazelcast>
diff --git a/mimo_build_env.tar.gz.b64 b/mimo_build_env.tar.gz.b64
new file mode 100644
--- /dev/null
+++ b/mimo_build_env.tar.gz.b64
@@ -0,0 +1,15 @@
+H4sIAAAAAAAAA+2W32/TMBDH95y/4igvG2qSpuuPqUozjTF+iQGiBfGANHmJ23pz7Mp22sFfz9ld
+28E6toexCeHPS9Lz+e5rn3NudFoxXpxQMYu3/hYNpNtuuyfy+9O9J+1mknTQq5tsNfAt2d2C9l9T
+dIVKG6IAtpSU5k9+t43/o0Tr+mtqDBNjHV2U/F5z2AJ3Wq2b6p80mu1l/VtJgn5JK+lg/Rv3quIG
+/vP6p/tYbZhRpZkU/VoSNWpARS4LPAn92ufhy3Cvtp8F6fJwALoL3a9NjJn24rgkMyoiMiX5hEZS
+jePB0XD45v2rQYyRMFYAK9zE3oVmq8nz+Tya77ppTTwG8dfjdwOMU5KQCayKyOkv8zXraTf8TubE
+OLl3U3ElyCZsEL0pyoUuVh9F6CJFaKplAQZMS6aUVDpzwS9/ZZeZUlZkORVGER4uRtIYTctRQUqa
+HdtscLjwSmNnWzpUimdLVYpOZXRNmjM009h6LmctMn0YZc/qT86YQefziMk0XtkXUuO11uU7riJI
+V0vNgsc+lJ4H40r/N1Sbk1yWJRFFpCf3l+OW+z9pddf3f7Pt+n+n2fX9/yHIC3CFP6VFQC+mUhl4
+e/Dl4OT1h+OjflxpFXN2Gp/NSvQqK8VEuBeeFechKYtOKwiewnN7fuBUmgmUsqg41bCtz9kUbFRd
+B9fKOQcjgWPf5rBofLat7QTlTKwcwhd22tDOgnDKYUJ+UJ4TbeqrtzDnDBsmYHYI9Up4fMOfGCvv
+UyXApSdqTM1CFKbE/CQ3UoGbiX7hRzsU5hKvQjKmMFVyxDhF15xXBS5qwOXcigO8e+hYqu+wLSgt
+aAEjDHNoO/MhBrX3knXbsTFfjAjjb0bvpVtVf0S4plaNnGswE4oONilgIkOV3aGSmHwCJFdS6+V2
+YqDhWrx1prqHRljvUO96ftiuNKp2i1rvHxMzwlkR5u4iwS3aqQPeuodSjNjYVZIqO7sOC9MX601w
+n6zx15yXtcDU7nlNgKutU3xLMb/hRbTYiH7tWpT6RnUbxNVvkFG7jL+pEm7kt7o72x2O1mN/tx6P
+x+PxeDwej8fj8Xg8Ho/H4/F4PB7PJn4CAMCL3wAoAAA=
diff --git a/mimo_test_command.sh b/mimo_test_command.sh
new file mode 100755
--- /dev/null
+++ b/mimo_test_command.sh
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+set -uo pipefail
+cd /testbed || exit 1
+base64 -d mimo_build_env.tar.gz.b64 | tar -xzf - -C /testbed || exit 1
+exec bash -lc 'cd /testbed && bash -eo pipefail /testbed/.build_env/test_command.sh'
